yax/refined/src/main/scala/doobie/refined/package.scala

```diff
@@ -0,0 +1,43 @@
+package doobie
+
+package object refined {
+
+  import eu.timepit.refined.api.{Refined, Validate}
+  import eu.timepit.refined.auto._
+  import eu.timepit.refined._
+  import scala.reflect.runtime.universe.TypeTag
+
+  import doobie.imports._
+  import doobie.util.invariant._
+
+  implicit def refinedMeta[T, P](implicit metaT: Meta[T], v: Validate[T, P], tag: TypeTag[T Refined P]): Meta[T Refined P] =
```

Comment:
Besides `Refined`, refined (the library) also supports `shapeless.tag.@@` and `scalaz.@@` as types for attaching a predicate to a base type. You could abstract over the tag type by adding another type parameter `F[_, _]` and an implicit parameter `rt: RefType[F]` which allows you to change the return type to `Meta[F[T, P]]`. In the implementation you can then use `rt.refine` and `rt.unwrap` for refining / unwrapping values. See for example [circe-refined](https://github.com/circe/circe/blob/master/modules/refined/shared/src/main/scala/io/circe/refined/package.scala) how you could use `RefType` here.

Btw, the same also applies to the `Composite` instance.